core/src/sigverify_shreds.rs

```diff
@@ -0,0 +1,395 @@
+    let mut slot_to_key_ix = HashMap::new();
+    for (i, (k, slots)) in keys.iter().enumerate() {
+        pubkeys.push(*k);
+        for s in slots {
+            slot_to_key_ix.insert(s, i);
+        }
+    }
+    let mut offsets = recycler_offsets.allocate("shred_offsets");
+    slots.iter().for_each(|packet_slots| {
+        packet_slots.iter().for_each(|slot| {
+            offsets.push((slot_to_key_ix.get(slot).unwrap() * size_of::<Pubkey>()) as u32);
+        });
+    });
+    //HACK: Pubkeys vector is passed along as a `Packets` buffer to the GPU
+    //TODO: GPU needs a more opaque interface, which can handle variable sized structures for data
+    //Pad the Pubkeys buffer such that it is bigger than a buffer of Packet sized elems
+    let num_in_packets =
+        (pubkeys.len() * size_of::<Pubkey>() + (size_of::<Packet>() - 1)) / size_of::<Packet>();
+    trace!("num_in_packets {}", num_in_packets);
+    //number of bytes missing
+    let missing = num_in_packets * size_of::<Packet>() - pubkeys.len() * size_of::<Pubkey>();
+    trace!("missing {}", missing);
+    //extra Pubkeys needed to fill the buffer
+    let extra = (missing + size_of::<Pubkey>() - 1) / size_of::<Pubkey>();
+    trace!("extra {}", extra);
+    trace!("pubkeys {}", pubkeys.len());
+    for _ in 0..extra {
+        pubkeys.push(Pubkey::default());
+        trace!("pubkeys {}", pubkeys.len());
+    }
+    trace!("pubkeys {:?}", pubkeys);
+    trace!("offsets {:?}", offsets);
+    (pubkeys, offsets)
+}
+
+fn shred_gpu_offsets(
+    mut pubkeys_end: u32,
+    batches: &[Packets],
+    recycler_offsets: &Recycler<TxOffset>,
+) -> (TxOffset, TxOffset, TxOffset, Vec<Vec<u32>>) {
+    let mut signature_offsets = recycler_offsets.allocate("shred_signatures");
+    let mut msg_start_offsets = recycler_offsets.allocate("shred_msg_starts");
+    let mut msg_sizes = recycler_offsets.allocate("shred_msg_sizes");
+    let mut v_sig_lens = vec![];
+    for batch in batches {
+        let mut sig_lens = Vec::new();
+        for packet in &batch.packets {
+            let sig_start = pubkeys_end;
+            let sig_end = sig_start + size_of::<Signature>() as u32;
+            let msg_start = sig_end;
+            let msg_end = sig_start + packet.meta.size as u32;
+            signature_offsets.push(sig_start);
+            msg_start_offsets.push(msg_start);
+            msg_sizes.push(msg_end - msg_start);
+            sig_lens.push(1);
+            pubkeys_end += size_of::<Packet>() as u32;
+        }
+        v_sig_lens.push(sig_lens);
+    }
+    (signature_offsets, msg_start_offsets, msg_sizes, v_sig_lens)
+}
+
+pub fn verify_shreds_gpu(
+    batches: &[Packets],
+    slot_leaders: &HashMap<u64, Pubkey>,
+    recycler_offsets: &Recycler<TxOffset>,
+    recycler_pubkeys: &Recycler<PinnedVec<Pubkey>>,
+    recycler_out: &Recycler<PinnedVec<u8>>,
+) -> Vec<Vec<u8>> {
+    let api = perf_libs::api();
+    assert!(api.is_some());
+    if api.is_none() {
```

Comment:
we should use a batched crossover point like we do in transaction sig_verify .

Reply:
TBD, after we bench it